tokio/src/sync/semaphore_ll.rs

```diff
@@ -640,209 +827,211 @@ impl ::std::error::Error for AcquireError {}
+    /// Only increments the number of permits *if* the waiter is currently
+    /// queued.
+    ///
+    /// # Returns
+    ///
+    /// `true` if the number of permits to acquire has been incremented. `false`
+    /// otherwise. On `false`, the caller should use `Semaphore::poll_acquire`.
+    fn try_inc_permits_to_acquire(&self, n: usize) -> bool {
+        let mut curr = WaiterState(self.state.load(Acquire));
 
-        self.acquire2()
-    }
+        loop {
+            if !curr.is_queued() {
+                assert_eq!(0, curr.permits_to_acquire());
+                return false;
+            }
 
-    fn acquire2(&self) -> Result<bool, AcquireError> {
-        use self::NodeState::*;
+            let mut next = curr;
+            next.set_permits_to_acquire(n + curr.permits_to_acquire());
 
-        match Idle.compare_exchange(&self.state, Assigned, AcqRel, Acquire) {
-            Ok(_) => Ok(true),
-            Err(Closed) => Err(AcquireError::closed()),
-            Err(_) => Ok(false),
+            match self.state.compare_exchange(curr.0, next.0, AcqRel, Acquire) {
+                Ok(_) => return true,
+                Err(actual) => curr = WaiterState(actual),
+            }
         }
     }
 
-    fn register(&self, cx: &mut Context<'_>) {
-        self.waker.register_by_ref(cx.waker())
-    }
+    /// Try to decrement the number of permits to acquire. This returns the
+    /// actual number of permits that were decremented. The delta betweeen `n`
+    /// and the return has been assigned to the permit and the caller must
+    /// assign these back to the semaphore.
+    fn try_dec_permits_to_acquire(&self, n: usize) -> usize {
+        let mut curr = WaiterState(self.state.load(Acquire));
 
-    /// Returns `true` if the permit has been acquired
-    fn cancel_interest(&self) -> bool {
-        use self::NodeState::*;
-
-        match Queued.compare_exchange(&self.state, QueuedWaiting, AcqRel, Acquire) {
-            // Successfully removed interest from the queued node. The permit
-            // has not been assigned to the node.
-            Ok(_) => false,
-            // The semaphore has been closed, there is no further action to
-            // take.
-            Err(Closed) => false,
-            // The permit has been assigned. It must be acquired in order to
-            // be released back to the semaphore.
-            Err(Assigned) => {
-                match self.acquire2() {
-                    Ok(true) => true,
-                    // Not a reachable state
-                    Ok(false) => panic!(),
-                    // The semaphore has been closed, no further action to take.
-                    Err(_) => false,
-                }
+        loop {
+            if !curr.is_queued() {
+                assert_eq!(0, curr.permits_to_acquire());
+            }
+
+            let delta = cmp::min(n, curr.permits_to_acquire());
+            let rem = curr.permits_to_acquire() - delta;
```

Comment:
Can the 2 lines be replaced with `let rem = curr.permits_to_acquire().saturating_sub(n)`?

Edit: Seems like we don't have the value for the return value in that case

Reply:
Yeah, we need to keep it.